### PR TITLE
GH-34484: [Substrait] add an option to disable augmented fields

### PR DESCRIPTION
**Pull request by EpsilonPrime**
### Rationale for this change

Augmented fields interfere with the schema passing between nodes.  When enabled they cause names/schema mismatching at the end of the plan.

### What changes are included in this PR?

Adds an option to disable augmented fields (defaulting to adding them), connects it everywhere it is called, and disables it in ReadRel conversion.

### Are these changes tested?

Yes.

### Are there any user-facing changes?

There are no API related changes however this will allow Substrait plans that consume local files to work without requiring a project/emit relation after the read relation to remove the unexpected fields.

* GitHub Issue: #34484

### CODE REVIEW

File: cpp/src/arrow/dataset/scanner.h [hunk 1]
```diff
@@ -287,10 +290,12 @@ struct ARROW_DS_EXPORT ProjectionDescr {
 
   /// \brief Create a default projection referencing fields in the dataset schema
   static Result<ProjectionDescr> FromNames(std::vector<std::string> names,
-                                           const Schema& dataset_schema);
+                                           const Schema& dataset_schema,
+                                           bool add_augmented_fields);
```
Comment (zeroshade) [inline cpp/src/arrow/dataset/scanner.h:294]: should we set a default value to ensure we don't break any existing consumers of this function?

File: cpp/src/arrow/dataset/scanner.h [hunk 2]
```diff
@@ -287,10 +290,12 @@ struct ARROW_DS_EXPORT ProjectionDescr {
...
 
   /// \brief Make a projection that projects every field in the dataset schema
-  static Result<ProjectionDescr> Default(const Schema& dataset_schema);
+  static Result<ProjectionDescr> Default(const Schema& dataset_schema,
+                                         bool add_augmented_fields);
```
Comment (zeroshade) [inline cpp/src/arrow/dataset/scanner.h:298]: same as above

**Review by zeroshade — COMMENTED** [review]
In general this looks fine to me, just a single nit pick from my end. But i'll leave final approval to someone more familiar with this section of the code.

File: cpp/src/arrow/dataset/scanner.h (hunk 2 above)
Comment (EpsilonPrime) [reply cpp/src/arrow/dataset/scanner.h]: Done.  Another alternative could be to pass through a ScanOptions data structure (still would require a default value though).

File: cpp/src/arrow/engine/substrait/serde_test.cc [hunk 3]
```diff
@@ -1064,6 +1064,86 @@ NamedTableProvider AlwaysProvideSameTable(std::shared_ptr<Table> table) {
...
+  ASSERT_STREQ(sink_node->kind_name(), "ConsumingSinkNode");
+  ASSERT_EQ(sink_node->num_inputs(), 1);
+  auto& prev_node = sink_node->inputs()[0];
+  ASSERT_STREQ(prev_node->kind_name(), "SourceNode");
+
+  plan->StartProducing();
+  ASSERT_FINISHES_OK(plan->finished());
+}
+
 TEST(Substrait, RelWithHint) {
```
Comment (westonpace) [inline cpp/src/arrow/engine/substrait/serde_test.cc:1147]: Minor nit: I think we have one or two spots in python where we have to do a column selection to workaround this issue.  We can probably remove these now.

e.g. https://github.com/apache/arrow/blob/main/python/pyarrow/tests/test_substrait.py#L93

**Review by westonpace — APPROVED** [review]
Awesome, thanks :)

**Review by zeroshade — APPROVED** [review]

**Pull request merged**